spec/std/spec/hooks_spec.cr

```diff
@@ -0,0 +1,108 @@
+require "../../spec_helper"
+
+describe Spec do
+  describe "hooks" do
+    it "runs in correct order" do
+      run(<<-CR).to_string.lines[..-5].should eq <<-OUT.lines
+        require "prelude"
+        require "spec"
+
+        before_all { puts "top:before_all" }
+        before_each { puts "top:before_each" }
```

Comment:
These blocks will be executed on all the specs of all the files, right?
If so, it is slightly missleading. I would expect them to run the specs of these file only.

With `Spec.before_each` it was more clear to me that the block will be executed beyond the scope of the file.

I'm not sure there is a clear way to solve this. Thoughts?

Reply:
Hm, I never considered file-local hooks. Not sure how useful they would be, especially considering that hooks that span multiple contexts would often be defind in helper files.
But I wouldn't be opposed to scope `before_each` in the root context to the current file and keep `Spec.before_each` as global.

But that would require to even introduce a file scope first. Maybe we don't need a separate instance for that but could use the location property of individual specs. But that feels like a separate story.

Reply:
Maybe a `__FILE__` argument can be used to filter in runtime if the block applies to the file or not.

And we can have a `before|after|around_suite` top level method for the things that in this PR are top

But from the syntax I would expect them to be per file.

The edge case would be if the `before_each` is used in a "support" file.

Reply:
Yeah, filtering by file shouldn't be an issue. It's just the question of *how* to design such a file context. Should it be a separate `Context` type? Or an implicit hidden `ExampleGroup`? Or do we just filter dynamically in `run_before_each`?

Reply:
`before_suite` could be an equivalent to `before_all` on the suite scope. But we would still need to differentiate `before_each` on suite and file scope.